app/ui/content/scripts/sessionrestore.js

```diff
@@ -0,0 +1,84 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+const SESSION_RESTORE_PAGE = 'tofino://sessionrestore';
+
+/**
+ * Upon loading tofino://sessionrestore, we check for params
+ * if this is used to inject history, or redirect to a page that was injected
+ * into history.
+ * Inspired by Firefox iOS's implementation:
+ * https://github.com/mozilla-mobile/firefox-ios/blob/75b22c3f03b07ede598293eae8bbddc241612224/Client/Assets/SessionRestore.html
+ */
+(function() {
+  const params = parseURLParams(window.document.location);
+
+  // If we have a `history` array and an `historyIndex`, then
+  // populate this tab's history with all the items in the array.
+  if (params.history) {
+    const historyList = JSON.parse(unescape(params.history));
+    const historyIndex = params.historyIndex ? Number(params.historyIndex) : historyList.length - 1;
+
+    // Push pages into `window.history`
+    for (let i = 0; i < historyList.length; i++) {
+      const url = getRestoreURL(historyList[i]);
+
+      // First, replace the session restore page (this page) with the first URL to be restored,
+      // or for additional entries, push into the state.
+      if (i === 0) {
+        history.replaceState({}, '', url);
+      } else {
+        history.pushState({}, '', url);
+      }
+    }
+
+    // Now we're on the last page pushed, but need to navigate to the selected history
+    // index in the params.
+    // `history.go` is relative to the current page, where 0 is the current page, -1 is back,
+    // 1 is forward, etc., whereas our historyIndex is relative to the 0th element in
+    // the history.
+    const relativeIndex = (historyIndex - history.length) + 1;
+    history.go(relativeIndex);
+
+    // Finally, reload the page to trigger the current history item's redirection, which
+    // will load the actual URL. We need to fire this using a preload script because we
+    // need to use the webview's refresh method, rather than anything in content, as that
+    // blows away all 'future' in the history list from the current page as it's considered
+    // navigation.
+    window._TOFINO_RELOAD();
```

Comment:
Changing the URL to be `tofino://historyrestore` and all related files, incase we have a sessionrestore page in the future (which would contain many historyrestore, for example)